lib/AST/GraphOperationInfo.cpp

```diff
@@ -73,6 +88,11 @@ void GraphOperationInfo::assertWithDump(bool cond,
 #endif // NDEBUG
 }
 
+StringRef GraphOperationInfo::getName() const {
```

Comment:
getName() is called repeatedly in some cases (e.g. GraphFunctionDeviceInfo::getForFunction()). It'd be more efficient (and also lead to cleaner code) if we do the parsing of the op name and all operands in the c'tor of GraphOperationInfo, and store them in (const) member fields.

This way GraphOperationInfo::decodeName() can just return references to those parsed objects. 

We may even be able to unify getName() and decodeName() into a single API, where returning the parsed "structured operands" could be optional.
(do you think reducing the public APIs of this class would make it easier to use? you can decide.)

Reply:
I like making the constructor parse everything, and making the API just return references to parsed objects.

How about two completely separate API methods `getOperationName()` and `getStructuredOperands()` that each just return one thing? This makes it very clear what you are getting, and makes it so that there is only one way to get each thing. I have updated the code to do this.

Reply:
SGTM